token-swap/program/src/processor.rs

```diff
@@ -414,16 +473,62 @@ impl Processor {
         let result = token_swap
             .swap_curve()
             .swap(
-                to_u128(amount_in)?,
+                to_u128(actual_amount_in)?,
                 to_u128(source_account.amount)?,
                 to_u128(dest_account.amount)?,
                 trade_direction,
                 token_swap.fees(),
             )
             .ok_or(SwapError::ZeroTradingTokens)?;
-        if result.destination_amount_swapped < to_u128(minimum_amount_out)? {
-            return Err(SwapError::ExceededSlippage.into());
-        }
+
+        // Re-calculate the source amount swapped based on what the curve says
+        let (source_transfer_amount, source_mint_decimals) = {
+            let source_amount_swapped = to_u64(result.source_amount_swapped)?;
+
+            let source_mint_data = source_token_mint_info.data.borrow();
+            let source_mint = Self::unpack_mint_with_extensions(
+                &source_mint_data,
+                source_token_mint_info.owner,
+                token_swap.token_program_id(),
+            )?;
+            let amount = if let Ok(transfer_fee_config) =
+                source_mint.get_extension::<TransferFeeConfig>()
+            {
+                source_amount_swapped.saturating_add(
+                    transfer_fee_config
+                        .calculate_epoch_fee_with_out(Clock::get()?.epoch, source_amount_swapped)
```

Comment:
I was initially confused by `calculate_epoch_fee_with_out` for a bit. I now understand what it is doing, but it feels more natural to me to have a function that computes the gross amount from the net amount than the fee from the net amount. It seems like what we ultimately need here is to compute the gross amount from the net amount as well.

I think the math could also be simplified a little bit, so let me meditate on this a little bit...